src/cockpit/channels/filesystem.py

```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
-                    if stat is not None:
-                        # copy permissions from existing file
-                        os.fchmod(fd, stat.st_mode)
-                        os.fchown(fd, stat.st_uid, stat.st_gid)
-                    break
-                except FileExistsError:
-                    continue
-                except PermissionError as exc:
-                    raise ChannelError('access-denied') from exc
-                except FileNotFoundError as exc:
-                    # directory of path does not exist
-                    raise ChannelError('not-found') from exc
-                except OSError as exc:
-                    raise ChannelError('internal-error', message=str(exc)) from exc
+                with contextlib.suppress(FileNotFoundError):  # delete is idempotent
+                    os.unlink(path)
             else:
-                raise ChannelError('internal-error',
-                                   message=f"Could not find unique file name for replacing {self._path}")
-
-            try:
-                self._tempfile = os.fdopen(fd, 'wb')
-            except OSError:
-                # Should Not Happen™, but let's be safe and avoid fd leak
-                os.close(fd)
-                self.unlink_temppath()
-                raise
-
-        self._tempfile.write(data)
-
-    def do_done(self):
-        if self._tempfile is None:
-            try:
-                os.unlink(self._path)
-            # crash on other errors, as they are unexpected
-            except FileNotFoundError:
-                pass
-        else:
-            self._tempfile.flush()
-
-            if self._tag and self._tag != tag_from_path(self._path):
-                raise ChannelError('change-conflict')
-
-            try:
-                os.rename(self._temppath, self._path)
-            # ensure to not leave the temp file behind
-            except FileNotFoundError as exc:
-                self.unlink_temppath()
-                raise ChannelError('not-found', message=str(exc)) from exc
-            except IsADirectoryError as exc:
-                self.unlink_temppath()
-                # not ideal, but the closest code we have
-                raise ChannelError('access-denied', message=str(exc)) from exc
-            except OSError as exc:
-                self.unlink_temppath()
-                raise ChannelError('internal-error', message=str(exc)) from exc
-
-            self._tempfile.close()
-            self._tempfile = None
+                # otherwise, spool data into a temporary file until EOF then rename into place...
+                with tempfile.NamedTemporaryFile(dir=dirname, prefix=f'.{basename}-', delete=False) as tmp:
+                    loop = asyncio.get_running_loop()
+                    while data is not None:
+                        await loop.run_in_executor(None, tmp.write, data)
+                        data = await self.read()
+
+                    os.fdatasync(tmp.fileno())
```

Comment:
Why is that necessary? The commit message doesn't give the "why".

Reply:
If we don't call fdatasync() then we're not guaranteed to have the new file's data on disk before the rename takes effect, and we could end up with a zero-length file.

This is how `g_file_set_contents()` does it, for example.  See this comment: https://gitlab.gnome.org/GNOME/glib/-/blob/704d0bb297cff6fc53beca48ccc2cfd061f0bf5e/glib/gfileutils.c#L1094

Reply:
Ah, so this isn't about rename() not working on "only in write cache" data (that would surprise me, as it's entirely an inode-level operation), but just generally making sure about data integrity, i.e. strong promises about that API ("if it succeeds, it's on disk"). This isn't documented. I don't mind adding it, it just wasn't clear why it's necessary for the renaming to work (... as it isn't). Thanks! (I'd prefer a second commit with an explanation, but not insisting)

Reply:
It's not yet clear to me that we accept this proposal PR in principle yet, so it would indeed be useful to have this if we stay with the old approach.

I'd feel pretty bad about calling that outside of a thread, though — it can block for seconds... and calling it in a thread in a way that works in callback-based code would start to get a bit hairy...